### PR TITLE
Eginez/public buckets

### PR DESCRIPTION
**Pull request by eginez**
Adds support for access types to objects in buckets. Tested by running the plan under docs/examples/storage/object

### CODE REVIEW

File: helpers_objectstorage.go [hunk 1]
```diff
@@ -28,6 +30,11 @@ var bucketSchema = map[string]*schema.Schema{
 		Required: true,
 		Computed: false,
 	},
+	"access_type": {
+		Type:     schema.TypeString,
+		Computed: false,
+		Optional: true,
```
Comment (josephholsten) [inline helpers_objectstorage.go]: probably want to add `Default: baremetal.NoPublicAccess,`

File: helpers_objectstorage.go [hunk 2]
```diff
@@ -28,6 +30,11 @@ var bucketSchema = map[string]*schema.Schema{
 		Required: true,
 		Computed: false,
 	},
+	"access_type": {
+		Type:     schema.TypeString,
+		Computed: false,
```
Comment (josephholsten) [inline helpers_objectstorage.go]: `Computed: false,` is the default. But it's included elsewhere, so at least it's consistent!

File: resource_obmcs_objectstorage_bucket.go [hunk 3]
```diff
@@ -86,6 +87,12 @@ func (s *BucketResourceCrud) Create() (e error) {
 		metadata := resourceObjectStorageMapToMetadata(rawMetadata.(map[string]interface{}))
 		opts.Metadata = metadata
 	}
+	if accessType, ok := s.D.GetOk("accessType"); ok {
```
Comment (josephholsten) [inline resource_obmcs_objectstorage_bucket.go]: it's usually better to skip these bits, depend on a default in the schema and end up with something like:

```go
opts.AccessType = baremetal.BucketAccessType(s.D.Get("accessType"))
```

File: resource_obmcs_objectstorage_bucket.go [hunk 4]
```diff
@@ -107,6 +114,12 @@ func (s *BucketResourceCrud) Update() (e error) {
 		opts.Metadata = metadata
 	}
 
+	if accessType, ok := s.D.GetOk("accessType"); ok {
```
Comment (josephholsten) [inline resource_obmcs_objectstorage_bucket.go]: meh, we really should extract this common data binding code from Create() & Update() someday.

File: resource_obmcs_objectstorage_object_test.go [hunk 5]
```diff
@@ -41,9 +41,13 @@ func (s *ResourceObjectstorageObjectTestSuite) SetupTest() {
 	s.TimeCreated = baremetal.Time{Time: time.Now()}
 
 	s.Config = `
+		variable "accesstype" {
+			default = "ObjectRead"
+		}
 		resource "baremetal_objectstorage_bucket" "t" {
 			compartment_id = "${var.compartment_id}"
 			name = "bucketID"
+			access_type="${var.accesstype}"
```
Comment (josephholsten) [inline resource_obmcs_objectstorage_object_test.go]: is there any reason not to just inline:
```hcl
access_type = "ObjectRead"
```

I also like to run my configs through `terraform fmt`, but clearly that wasn't the case when you came to the code.

File: resource_obmcs_objectstorage_bucket.go (hunk 3 above)
Comment (josephholsten) [inline resource_obmcs_objectstorage_bucket.go]: Woah! this absolutely needs to be using key `"access_type"` not `"accessType"`.

File: helpers_objectstorage.go (hunk 1 above)
Comment (josephholsten) [inline helpers_objectstorage.go]: oh, while we're at it we probably want to add something like `ValidationFunc: validation.StringInSlice([]string{baremetal.NoPublicAccess, baremetal.ObjectRead}, true)`, see https://godoc.org/github.com/hashicorp/terraform/helper/validation#StringInSlice

File: resource_obmcs_objectstorage_bucket.go (hunk 3 above)
Comment (eginez) [inline resource_obmcs_objectstorage_bucket.go]: nice catch, last minute refactoring! :/

File: helpers_objectstorage.go (hunk 1 above)
Comment (eginez) [inline helpers_objectstorage.go]: this is great, thanks for the pointer

**Comment by codycushing** [comment]
Can you update bucket.md with the new param

File: docs/examples/storage/object/README.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
+    #     ___  ____     _    ____ _     _____
+    #    / _ \|  _ \   / \  / ___| |   | ____|
+    #   | | | | |_) | / _ \| |   | |   |  _|
+    #   | |_| |  _ < / ___ | |___| |___| |___
+    #    \___/|_| \_/_/   \_\____|_____|_____|
+***
+## Creating buckets
+This example creates a single bucket inside a comparment
```
Comment (codycushing) [inline docs/examples/storage/object/README.md]: comparment -> compartment

File: docs/examples/storage/object/variables.tf [hunk 7]
```diff
@@ -0,0 +1,11 @@
+variable "tenancy_ocid" {}
+variable "user_ocid" {}
+variable "fingerprint" {}
+variable "private_key_path" {}
+variable "namespace_name" {}
+variable "compartment_ocid" {}
+
+variable "BucketName" {
+    default = "BucketOne"
+}
```
Comment (codycushing) [inline docs/examples/storage/object/variables.tf]: Defining this variable only sort of complicates the example, guessing it was just a test

**Comment by eginez** [comment]
Thanks for the input Cody, latest commit should have your comments addressed

**Comment by eginez** [comment]
wohooo!

**Pull request merged**